[Fastlane.Swift] Fix `OptionalConfigValue` for `Any`-based types.

minuscorp opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
`OptionalConfigValue` does not provide an `ExpressibleBy` interface for `Any?` parameters, so they have to be filtered out afterwards.
Fixes #18654

### Description
`OptionalConfigItem` does not apply to `Any` types to avoid a breaking compatibility against past releases. This PR resolves the issue.

### Testing Steps
1. Install the following branch:
```ruby
gem 'fastlane', :branch => 'https://github.com/minuscorp/fastlane/fix-configvalue-with-any-type'
```
2. Use an action with an `Any` value:
```swift
incrementBuildNumber(buildNumber: <#T##Any?#>)
```
3. Check that the value is being passed when it is not `nil`.
4. 🚀

minuscorp committed:
[Fastlane.Swift] Fix `OptionalConfigValue` for `Any`-based types.

ainame commented on fastlane/lib/fastlane/swift_fastlane_function.rb:
```diff
@@ -325,7 +326,7 @@ def implementation
       else
         implm += "let args = [#{args.group_by { |h| h[:name] }.keys.join(",\n")}]\n"
-        implm += ".compactMap { $0 }\n"
+        implm += ".compactMap { $0?.nilify(when: { $0.value == nil }) }\n"
```
```suggestion
        implm += ".filter { $0?.value != nil } \n"
```

What about this? Isn't this simpler?

minuscorp replied:
🤦‍♂️ I'm an idiot

minuscorp replied:
Wasn't exactly doing that, but gave me the idea to get rid of the `nilify` method ❤️

ainame reviewed (approved):
Had a minor suggestion to make it simpler but LGTM

minuscorp committed:
Review comments

joshdholtz reviewed (changes requested):
I tried doing `incrementBuildNumber(buildNumber: 5)` but I got 👇 

```sh
value of protocol type 'Any' cannot conform to 'ExpressibleByIntegerLiteral'; only struct/enum/class types can conform to protocols
```

Am I doing something wrong? 😇

minuscorp commented:
But it works fine with you for float and string values?🤔

minuscorp committed:
Fix for `Any?` with `ExpressibleByIntegerLiteral`

RyuX51 commented:
I tried the `fix-configvalue-with-any-types` branch and I'm getting

```
▸ Compiling Fastfile.swift

❌  .../Fastfile.swift:22:21: cannot convert value of type 'String' to expected argument type 'OptionalConfigValue<String?>'

      echo(message: "Unknown parameter value \(bumpType)")
                    ^


** BUILD FAILED **
```

for the following code:
```
  func betaLane(withOptions options: [String: String]?) {

    let bumpType = options?["bumpType"]

    if !["major", "minor", "patch"].contains(bumpType) {
      echo(message: "Unknown parameter value \(bumpType)")
      return
    }
```

If I do without string interpolation it works fine. Thanks.

Since I'm coming from [18654](https://github.com/fastlane/fastlane/issues/18654), I'll just drop this here.

minuscorp committed:
Added support for String interpolation

minuscorp commented:
@RyuX51 could you check that your issue is solved now?

RyuX51 commented:
@minuscorp I'm sorry, `c1b9cf1`still behaves the same.

minuscorp commented:
`bumpType` is `String` or `String?`?

RyuX51 commented:
It's `String`, but this isn't the problem. For example

```echo(message: "Unknown " + "parameter")```

and

```echo(message: "Unknown \("parameter")")```

also lead to the same error whereas

```echo(message: "Unknown parameter")```

is just fine.

minuscorp commented:
I just tested using 

```swift
let string = "Foo"

echo("Message \(string)")
```
Working🤔

RyuX51 commented:
@minuscorp

```Using fastlane 2.182.0 from https://github.com/minuscorp/fastlane (at fix-configvalue-with-any-types@c1b9cf1)```

So this should be the right commit, right?

```
▸ Compiling Fastfile.swift

❌  /Users/mario/Repository/VJeton/fastlane/Fastfile.swift:24:12: missing argument label 'message:' in call

      echo("Message \(string)")
           ^



❌  /Users/mario/Repository/VJeton/fastlane/Fastfile.swift:24:12: cannot convert value of type 'String' to expected argument type 'OptionalConfigValue<String?>'

      echo("Message \(string)")
           ^


** BUILD FAILED **
```

And when I add the `message:` it's again:

```
▸ Compiling Fastfile.swift

❌  /Users/mario/Repository/VJeton/fastlane/Fastfile.swift:24:21: cannot convert value of type 'String' to expected argument type 'OptionalConfigValue<String?>'

      echo(message: "Message \(string)")
                    ^


** BUILD FAILED **
```

minuscorp commented:
Pretty confusing to me 😔 
![image](https://user-images.githubusercontent.com/3819883/118326508-5aeede80-b505-11eb-801f-0619a69556f7.png)

RyuX51 commented:
@minuscorp Even `echo(message: String("Message"))` fails for me. 🤷‍♂️

minuscorp commented:
I'd reset the repo and try to run the `generate_swift_api` lane before going further, maybe something in the project it's corrupted😞

minuscorp commented:
@joshdholtz Can you replicate the issue or it is working fine for you?🤔

minuscorp commented:
Any news @RyuX51 regarding the issue with interpolation?

RyuX51 commented:
@minuscorp Well, I haven't investigated further. I just set up a new project where I wanted to try a Fastlane.swift for the first time. The fact that I get this error while wanting to show a value in an informational output is not really a blocker for me. This problem just sounded a lot like what you are fixing here is related, and I just wanted to let you know. Since you can't even replicate the bug, I don't want to concern you with that.

joshdholtz commented:
@minuscorp I’m still getting…

```sh
value of protocol type 'Any' cannot conform to 'ExpressibleByUnicodeScalarLiteral'; only struct/enum/class types can conform to protocols
```

when my `Fastlane.swift` file looks like…

```swift
incrementBuildNumber(buildNumber: "5")
```

Are you not getting that same thing? 🤔

minuscorp committed:
Fix for optional meta values

minuscorp commented:
@joshdholtz indeed, I removed all conformances to `Optional`,  as it seems to not accept `Any` meta value, differently from `AnyObject` which it could conform to those extensions. Fixed and tested:

```swift
incrementBuildNumber(buildNumber: 1.0, xcodeproj: "asd")
incrementBuildNumber(buildNumber: "1.0")
incrementBuildNumber(buildNumber: 1)
```

joshdholtz reviewed (approved):
This works for me! ❤️ LGTM

Merge whenever you are ready
